tests/queries/0_stateless/01622_join_implicit_cast.sql

```diff
@@ -0,0 +1,174 @@
+DROP TABLE IF EXISTS t1;
+DROP TABLE IF EXISTS t2;
+
+CREATE TABLE t1 (a UInt8, b UInt16) ENGINE = TinyLog;
+CREATE TABLE t2 (a String, b Nullable(Int64)) ENGINE = TinyLog;
+
+INSERT INTO t1
+SELECT number + 2 as a, 100 + number + 2 as b FROM system.numbers LIMIT 1, 15;
+
+INSERT INTO t2
+SELECT number as a, 200 + number as b FROM system.numbers LIMIT 1, 10;
+INSERT INTO t2 VALUES ('-1', NULL);
+
+SET join_algorithm = 'hash';
+
+SELECT '--- hash ---';
+
+SELECT '- full -';
+SELECT * FROM t1 FULL JOIN t2 ON t1.a == t2.a ORDER BY (toInt32(if(empty(t2.a), '0', t2.a)), toInt32(t1.a));
```

Comment:
I thought that implicit conversion of ints to string might be too much, but looks like our `equals` works just the same way:
```
:) select toUInt8(1) == '1'

┌─equals(toUInt8(1), '1')─┐
│                       1 │
└─────────────────────────┘
```

Is this patch fully consistent with `equals`, e.g. we'll be able to use the same equalities for equi-join that the `equals` supports?

Reply:
In the initial feature request, Alexey only mentions `getLeastSupertype`-like logic, converting between different nullability or int types, w/o complex things like string parsing etc. So you're doing a bit more. But I think any behavior consistent with `equals` makes sense here.

Reply:
It both uses `convertFieldToType`, so I suppose behavior is consistent